File: python_kata/src/trees/binary_tree/bfs_tree_depth.py
from trees.binary_tree.node import BinaryTree
from data_structures.queue import Queue

def tree_depth(tree: BinaryTree[int]) -> int:
    if not tree.root:
        return -1

    maxDepth = 0
    cache = {}
    q = Queue()

    q.enqueue(tree.root)
    cache[tree.root, 0]

    while q.length:
        node = q.deque()
        if not node:
            continue

        depth = cache[node]
        maxDepth = depth if depth > maxDepth else maxDepth
        
        if node.left:
            q.enqueue(node.left)
            cache[node.left] = depth + 1
        if node.right:
            q.enqueue(node.right)
            cache[node.right] = depth + 1

    return maxDepth

File: go-kata/internal/trees/binary_tree/bfs_tree_depth.go
package binary_tree

func BfsTreeDepth(tree *BinaryTree[int]) int {
	if tree.Root == nil {
		return -1
	}

	maxDepth := 0
	q := []*BinaryNode[int]{}
	cache := make(map[*BinaryNode[int]]int)

	q = append(q, tree.Root)
	cache[tree.Root] = 0

	for len(q) > 0 {
		node := q[0]
		if node == nil {
			continue
		}

		depth := cache[node]

		if depth > maxDepth {
			maxDepth = depth
		}

		if node.Left != nil {
			q = append(q, node.Left)
			cache[node.Left] = depth + 1
		}

		if node.Right != nil {
			q = append(q, node.Right)
			cache[node.Right] = depth + 1
		}
	}

	return maxDepth
}
